cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -141,6 +142,285 @@ void inplace_bitmask_binop(
   stream.synchronize();
 }
 
+namespace {
+
+/**
+ * For each range `[first_bit_indices[i], last_bit_indices[i])`
+ * (where 0 <= i < `num_ranges`), count the number of bits set outside the range
+ * in the boundary words (i.e. words that include either
+ * `first_bit_indices[i]'th` bit or `(last_bit_indices[i] - 1)'th` bit) and
+ * subtract the count from the range's null count.
+ *
+ * Expects `0 <= first_bit_indices[i] <= last_bit_indices[i]`.
+ *
+ * @param[in] bitmask The bitmask whose non-zero bits outside the range in the
+ * boundary words will be counted.
+ * @param[in] num_ranges The number of ranges
+ * @param[in] first_bit_indices The indices (inclusive) of the first bit in each
+ * range
+ * @param[in] last_bit_indices The indices (exclusive) of the last bit in each
+ * range
+ * @param[in,out] null_counts The number of non-zero bits in each range to be
+ * updated
```

Comment:
Sorry that I'm a bit OCD about the style. This is just a suggestion (align multiple-line param doxygen with the parameter name), you can ignore it if you don't like :smiley: 
```suggestion
 * @param[in] bitmask The bitmask whose non-zero bits outside the range in the
 *                    boundary words will be counted.
 * @param[in] num_ranges The number of ranges
 * @param[in] first_bit_indices The indices (inclusive) of the first bit in each
 *                     range
 * @param[in] last_bit_indices The indices (exclusive) of the last bit in each
 *                     range
 * @param[in,out] null_counts The number of non-zero bits in each range to be
 *                          updated
```